### ISSUE TITLE
Introduce 'computed' column type for secondary indexes token column

### ISSUE THREAD

**Issue by psarna** [scylladb/scylla#3512]
Currently issue #3423 is solved by letting materialized views know that they are used as a back-end for secondary indexes. Then, special token column can be computed on updates to ensure token order on queries, but at the same time this implementation violates layering - ideally materialized views should be used by secondary indexes without having any special cases for them.

Alternative solution is to introduce a `computed` column type, which cannot be explicitly updated, but is instead generated from other columns on every update. Token column for secondary indexes can then be implemented as a `computed` column.

Discussion link: https://groups.google.com/forum/#!searchin/scylladb-dev/token$20column%7Csort:date/scylladb-dev/19N_-WPkl04/ZrtiuzLKBQAJ

**Comment by nyh** [comment]
We don't even need this to be a general "computed" column, why can't this be a boolean flag as today, just set from the SI layer? I.e., SI layer asks the MV layer to enable this flag - not the other way around?

**Comment by psarna** [comment]
Right now the only use case is token column and it's a bunch of `if` statements in view code, the question is if `computed` column might be useful somewhere else in the future. Because then it would make sense for it to be a first class column citizen.

Also, if `computed` was a column type stuck to a column definition, all these `if`s would get much shorter and most of the tests would hide behind accessors like `get_regular_columns()` etc.

Finally (that's my pure subjective opinion), in the and it *is* a special column type, so it feels natural to have it listed as one of the column types, if nothing else it makes all the code self-descriptive.

**Comment by duarten** [comment]
Another advantage of having an explicit computed column type is that we can leverage it to encode transformations on a base column.

**Comment by psarna** [comment]
I started working on a POC for this one and I stumbled upon one little issue. Currently in materialized views we just check if the view is actually backing a secondary index, and if so, a special `token` column is handled. Now, if we switch to be able to check `column.is_computed()` this won't be necessary and SI code will be decoupled from MV, which is good, but it's also a backward compatibility issue.

In current release the `token` column will be present as a regular column in `system_schema`. If we introduce computed columns and keep them in `system_schema.computed_columns` (following view virtual columns' suit), it will not work correctly in a rolling upgrade - for indexes created in current version the column will be listed as regular, while the future code will expect them to be marked as computed from `system_schema.computed_columns`, which will result in indexes not working properly.

Option 1 to deal with it is to create a routine (script) that fixes schema tables before upgrading, but that's prone to errors and I'm not a fan.
Option 2 is performing an additional check if the column should be marked as `computed` when reading schema tables, which will make loading interfaces more ugly (because base schema for each view would need to be available to perform the check). Also, I'd need to make sure if `database.cc::parse_system_tables` is the only place when schema tables are parsed and schemas added - I think it is, but can somebody confirm?
Option 3 is performing this check in `view_ptr::initialize_base_dependent_fields`, which sounds like a nice place to do the check, but by then the schema is already created and current interfaces do not really allow modifying column definitions - all currently available iterators offer `const` access only - that can of course be changed, but having a mutable `_is_computed` attribute or exposing new `non-const` accessors just because of backward compatibility seems fishy.

So, to sum up, option 2 would be best, right? Unless somebody has another idea how to handle this.

**Comment by duarten** [comment]
An alternative is to have `system_schema.computed_columns` just mark which columns in the other columns table are computed columns. Essentially, you'd need to read from both system tables to create a computed column. 

The order of computed columns matters, because they can be key columns. This is unlike virtual columns, which just appear at the end as regular columns. So maybe the above is the best solution?

The other alternative is to have `schema_tables` support both options. When loading a secondary index, it looks for the token column in both places. Eventually, when we stop supporting versions that create indexes in the old way, we'll drop this code. (We also need a feature flag to prevent new nodes from creating index schemas in the new way while the cluster is being upgraded.)

**Comment by nyh** [comment]
On Wed, Mar 13, 2019 at 5:30 PM Duarte Nunes <notifications@github.com>
wrote:

> An alternative is to have system_schema.computed_columns just mark which
> columns in the other columns table are computed columns. Essentially, you'd
> need to read from both system tables to create a computed column.
>
> The order of computed columns matters, because they can be key columns.
> This is unlike virtual columns, which just appear at the end as regular
> columns. So maybe the above is the best solution?
>
Also note that the reason why the computed columns had to appear only in a
separate table, and not in the original columns table, was because drivers
read the columns table to figure out which columns exist, and we don't want
them to know about the virtual columns. But if you want drivers to be able
to see the new "computer columns", they need to be listed in the normal
columns table, not just in a new one.

For secondary indexes this probably doesn't matter (drivers will never try
to inquire about the columns of a secondary index's view, which the user
shouldn't even know exists), but for other future uses of computed columns,
this may make a difference.

**Comment by nyh** [comment]
@psarna this should be marked done, right?
Via your patch series db8d4a0cc649c74e30198dd83b082b0ad50b1314 and specifically commit a8f7d64a08f029f3355d9540f310a4816ca70154 ?

**Comment by psarna** [comment]
Yes, it's all done.
